ls1intum/Artemis
# `Discussion`: Display names of reacting users when hovering a reaction

### Checklist
#### General



#### Client

### Motivation and Context


1. Proposed by a user requesting the feature in #4306.
2. The PR also fixes a bug described in #4323.
3. It fixes the window title for the course discussion page that got lost when restructuring module loading to reduce the bundle size (see pictures)

### Description

For the features that shown the names of reacting users, we make use of the existing functionality in `buildReactionMetaDataMap` that collects reacting users for a each reaction on a postings and determines if the logged in user is among them (in this case the emoji is highlighted). We transform this information with a newly introduced pipe `reactingUsersOnPosting` that takes the array of reacting users and outputs a suitable string taking into account several aspects:

- if only the currently logged in student reacted -> "you"
- if there are only two users reacting their names are joined with "and" (or "und") -> "x and y reacted."
- if there are more than two users reacting their names are joined with ", " but the last name is appended with "and" -> (x, y and z reacted"
- if the logged in user reacted, he/she is included in the list of reacting users by having "You" (or "Du") in the beginning of the string -> "you, x, y and z reacted".
- if there are more than 10 (e.g. 12)  people reacting, the list is trimmed -> "a, b, c, d, e, f, g, h, i, j and 2 other users reacted."

### Steps for Testing

Prerequisites:
- 1 User
- 1 Course with postings enabled

1. Log in to Artemis
2. Navigate to course discussion overview -> name in tab should be `Discussion` or `Diskussion`
3. Hover over reactions -> check that the showed tooltips on reacting users follows the rules introduced in the description above
4. Change the language -> check if the tooltip language automatically changed accordingly, without reloading the page
5. Add a reaction -> check that the tooltip on reacting users is updated correctly
6. Remove a reaction -> check that the tooltip on reacting users is updated correctly
7. For reactions with more than 10 users, check that the tooltip on reacting users is trimmed correctly (you can use this posting to test, it already has 11 reacting users: https://artemistest.ase.in.tum.de/courses/477/lectures/184?postId=1798)
8. **As windows user** confirm that when hovering the button to add a reaction the scrollbar of the window does not flicker anymore and you can click the button without any disturbances.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





`reacting-users-on-posting.pipe.ts`: 100%
`posting-reaction-bar.component.ts`: 85,7%
`post-reaction-bar.compnent.ts`: 100%
`answer-post-reaction-bar.component.ts`: 95,1%

### Screenshots


#### 1. Hovering over reactions to see user names (English and German):

https://user-images.githubusercontent.com/41366522/144131513-9f4dfd28-1ff0-4f51-9baf-0a49f2e2e382.mov

#### 2. Changed tooltip for choosing a reaction -> hint is now encapsulated in a button "title" which looks a bit different than the tooltips but this prevents the scrollbar from flickering in windows (#4323) which prevents the button from being usable.
**Before:**
<img width="205" alt="Bildschirmfoto 2021-11-30 um 13 01 33" src="https://user-images.githubusercontent.com/41366522/144044545-2969b806-f30e-4bf2-ac2a-da1498f15da8.png">

**After:**
<img width="182" alt="Bildschirmfoto 2021-11-30 um 13 01 58" src="https://user-images.githubusercontent.com/41366522/144045802-6963704a-3ca3-4bb2-aa4f-fe12cb6c37ff.png">

#### 3. Fixed window title for course discussion page (before: red, after: green):
<img width="1391" alt="Bildschirmfoto 2021-11-30 um 12 56 55" src="https://user-images.githubusercontent.com/41366522/144044180-b764a995-8988-4265-be40-0a51d372c7fb.png">
